File: 2025/four.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

static int ROW_COUNT = 140;
static int ROW_LENGTH = 140;

void run_four(FILE* input) {
    char rows[ROW_COUNT + 2][ROW_LENGTH + 1];

    int rowNum = 1;
    while (fgets(rows[rowNum]+1, ROW_LENGTH+1, input) != NULL) {
        rows[rowNum][0] = '.';
        rows[rowNum][ROW_LENGTH] = '.';

        rowNum++;
    }

    printf("Size: %d x %d\n", ROW_LENGTH, ROW_COUNT);

    for (int i = 0; i < ROW_LENGTH + 1; i++) {
        rows[0][i] = '.';
        rows[ROW_COUNT+1][i] = '.';
    }

    int reachableCount = 0;
    for (int y = 1; y < ROW_COUNT; y++) {
        for (int x = 1; x < ROW_LENGTH; x++) {
            char value = rows[y][x];
            if (value != '@') {
                putchar(value);
                continue;
            }

            int count = 0;

            // Previous Row
            if (rows[y - 1][x - 1] == '@') {
                count++;
            }

            if (rows[y][x - 1] == '@') {
                count++;
            }

            if (rows[y + 1][x - 1] == '@') {
                count++;
            }

            // Same Row
            if (rows[y - 1][x] == '@') {
                count++;
            }

            if (rows[y + 1][x] == '@') {
                count++;
            }

            // Next Row
            if (rows[y - 1][x + 1] == '@') {
                count++;
            }

            if (rows[y][x + 1] == '@') {
                count++;
            }

            if (rows[y + 1][x + 1] == '@') {
                count++;
            }

            if (count < 4) {
                putchar('x');
                reachableCount++;
            }
            else {
                putchar('@');
            }

        }

        putchar('\n');
    }

    printf("Reachable: %d\n", reachableCount);
}

File: 2025/four.pl
#!/usr/bin/perl

use strict;
use warnings;

my $rows;
while (my $line = <STDIN>) {
    chomp($line);
    my @row = split(//, $line);
    unshift(@row, '.');
    push(@row, '.');

    push(@$rows, \@row);
}

my $width = scalar(@{@$rows[0]}) - 2;
my $height = scalar(@$rows);
print "Size: $width x $height\n";

my @empty_row = ('.') x ($width + 2);
unshift(@$rows, \@empty_row);
push(@$rows, \@empty_row);

my $reachable_count = 0;
foreach my $x (1 .. $width) {
    foreach my $y (1 .. $height) {
        my $value = $$rows[$x][$y];
        if ($value ne '@') {
            print $value;
            next;
        }

        my $count = 0;

        # Previous Row
        if ($$rows[$x - 1][$y - 1] eq '@') {
            $count++;
        }

        if ($$rows[$x][$y - 1] eq '@') {
            $count++;
        }

        if ($$rows[$x + 1][$y - 1] eq '@') {
            $count++;
        }

        # Same Row
        if ($$rows[$x - 1][$y] eq '@') {
            $count++;
        }

        if ($$rows[$x + 1][$y] eq '@') {
            $count++;
        }

        # Next Row
        if ($$rows[$x - 1][$y + 1] eq '@') {
            $count++;
        }

        if ($$rows[$x][$y + 1] eq '@') {
            $count++;
        }

        if ($$rows[$x + 1][$y + 1] eq '@') {
            $count++;
        }

        if ($count < 4) {
            print 'x';
            $reachable_count++;
        }
        else {
            print '@';
        }        
    }

    print "\n";
}

print "Reachable: $reachable_count\n";